### PR TITLE
[FLINK-34903][MySQL][Feature] Add mysql-pipeline-connector with tables.exclude option to exclude unnecessary tables

### PR DESCRIPTION
**Pull request by shiyiky**
desc :[FLINK-34903](https://issues.apache.org/jira/browse/FLINK-34903)

### CODE REVIEW

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/org/apache/flink/cdc/connectors/mysql/factory/MySqlDataSourceFactory.java [hunk 1]
```diff
@@ -93,6 +96,7 @@ public DataSource createDataSource(Context context) {
         String username = config.get(USERNAME);
         String password = config.get(PASSWORD);
         String tables = config.get(TABLES);
+        String tableExcludeList = config.get(TABLE_EXCLUDE_LIST);
```
Comment (laglangyue) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/org/apache/flink/cdc/connectors/mysql/factory/MySqlDataSourceFactory.java:99]: source.`tables` include more than one table , so it should be `tables.exclude`?

File: docs/content.zh/docs/connectors/mysql.md [hunk 2]
```diff
@@ -202,6 +202,15 @@ pipeline:
       <td>Integer</td>
       <td>连接池大小。</td>
     </tr>
+    <tr>
+      <td>table.exclude.list</td>
```
Comment (laglangyue) [inline docs/content.zh/docs/connectors/mysql.md:206]: source.`tables` include more than one table , so it should be `tables.exclude`?

**Review by laglangyue — APPROVED** [review]
LGTM, as a volunteer.

**Comment by Jiabao-Sun** [comment]
Thanks @shiyiky for this contribution.
@ruanhang1993, could you help review this when you have time?

File: docs/content/docs/connectors/mysql.md [hunk 3]
```diff
@@ -205,6 +205,16 @@ pipeline:
           <td>Integer</td>
           <td>The connection pool size.</td>
     </tr>
+    <tr>
+      <td>table.exclude.list</td>
```
Comment (PatrickRen) [inline docs/content/docs/connectors/mysql.md:209]: TBH I'm not in favor of adding this option. It reminds me the nightmare using `maven-shade-plugin` that also [supports both include and exclude](https://maven.apache.org/plugins/maven-shade-plugin/shade-mojo.html), which always make me confused about the priority of these two options. [Debezium](https://debezium.io/documentation/reference/2.6/connectors/mysql.html#mysql-property-table-exclude-list) also doesn't support setting both inclusion and exclusion together. 

The `tables` option uses regex for matching the table id. Considering regex itself supports negation, what about controlling the exclusion list using the negation in regex?

**Review by PatrickRen — COMMENTED** [review]
@shiyiky Thanks for the PR! I left my concern about introducing this new option.

File: docs/content/docs/connectors/mysql.md (hunk 3 above)
Comment (shiyiky) [reply docs/content/docs/connectors/mysql.md:209]: 我赞同你的观点，但是大部分用户在进行整库通的过程中，他们用db.*来进行所有表数据的采集，然后希望排除部分表，用户希望的是简单复杂，用户写正则来匹配表，可能这对部分来说来说不如排除表来的简单。
我参考了部分基于cdc连接器的改造整库同步，他们都会有排除表这个选项，例如阿里云的cdas，doris的整库同步，paimon的整库同步，可能这个选项‘table.exclude.list’ 和debezium的参数保持太一致不太好，或许更名其他名字来区分二义性。

Comment (shiyiky) [reply docs/content/docs/connectors/mysql.md:209]: 我赞同你的观点，但是大部分用户在进行整库通的过程中，他们用db.*来进行所有表数据的采集，然后希望排除部分表，用户希望的是简单复杂，用户写正则来匹配表，可能这对部分来说来说不如排除表来的简单。
我参考了部分基于cdc连接器的改造整库同步，他们都会有排除表这个选项，例如阿里云的cdas，doris的整库同步，paimon的整库同步，可能这个选项‘table.exclude.list’ 和debezium的参数保持太一致不太好，或许更名其他名字来区分二义性。


doris's  link :https://doris.apache.org/docs/ecosystem/flink-doris-connector#use-flinkcdc-to-access-multiple-tables-or-the-entire-database-supports-mysql-oracle-postgresql-sqlserver
image

paimon's link :https://paimon.apache.org/docs/0.7/cdc-ingestion/mysql-cdc/

Comment (shiyiky) [reply docs/content/docs/connectors/mysql.md:209]: 我赞同你的观点，但是大部分用户在进行整库通的过程中，他们用db.*来进行所有表数据的采集，然后希望排除部分表，用户希望的是简单复杂，用户写正则来匹配表，可能这对部分来说来说不如排除表来的简单。
我参考了部分基于cdc连接器的改造整库同步，他们都会有排除表这个选项，例如阿里云的cdas，doris的整库同步，paimon的整库同步，可能这个选项‘table.exclude.list’ 和debezium的参数保持太一致不太好，或许更名其他名字来区分二义性。

doris's link :https://doris.apache.org/docs/ecosystem/flink-doris-connector#use-flinkcdc-to-access-multiple-tables-or-the-entire-database-supports-mysql-oracle-postgresql-sqlserver


paimon's link :https://paimon.apache.org/docs/0.7/cdc-ingestion/mysql-cdc/

Comment (shiyiky) [reply docs/content/docs/connectors/mysql.md:209]: cdas：
 
image

doris：
image

Comment (PatrickRen) [reply docs/content/docs/connectors/mysql.md:209]: Thanks for the input. With these examples I think it makes sense to support exclusions. 

About the name of the config: what about `tables.exclude`? The value is not a list but a regex, and using plural form is to align with the option `tables`.

And we need to explain to users about how we apply these filters: match using `tables` first, then drop tables matches `tables.exclude`.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/test/java/org/apache/flink/cdc/connectors/mysql/source/MySqlDataSourceFactoryTest.java [hunk 4]
```diff
@@ -79,6 +80,49 @@ public void testNoMatchedTable() {
...
+    public void testExcludeTable() {
+        inventoryDatabase.createAndInitialize();
+        Map<String, String> options = new HashMap<>();
+        options.put(HOSTNAME.key(), MYSQL_CONTAINER.getHost());
+        options.put(PORT.key(), String.valueOf(MYSQL_CONTAINER.getDatabasePort()));
+        options.put(USERNAME.key(), TEST_USER);
+        options.put(PASSWORD.key(), TEST_PASSWORD);
+        // db has three tables , table.list= (products,orders shipments)
+        options.put(TABLES.key(), inventoryDatabase.getDatabaseName() + ".prod\\.*");
+        String tableExcludeList = inventoryDatabase.getDatabaseName() + ".prod\\.orders";
```
Comment (PatrickRen) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/test/java/org/apache/flink/cdc/connectors/mysql/source/MySqlDataSourceFactoryTest.java:93]: Here the inclusion pattern is `prod.*`, which matches only one table `products`. The exclusion pattern is `prod.orders`, which doesn't match any table, so I'm afraid this case doesn't truly test if the exclusion works as expected. 

What about using `\\.*` for matching all tables in inclusion, and `orders` for exclusion? The expected output should be `products` and `shipments` after applying the filter.

**Review by PatrickRen — COMMENTED** [review]
@shiyiky Thanks for the explanation! I left some comments.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/test/java/org/apache/flink/cdc/connectors/mysql/source/MySqlDataSourceFactoryTest.java (hunk 4 above)
Comment (shiyiky) [reply flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/test/java/org/apache/flink/cdc/connectors/mysql/source/MySqlDataSourceFactoryTest.java:93]: u are right and  I will adapter it，tks。

**Comment by shiyiky** [comment]
@PatrickRen  I have done it .Please review it if u have free time ,tks.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/org/apache/flink/cdc/connectors/mysql/source/MySqlDataSourceOptions.java [hunk 5]
```diff
@@ -231,4 +231,16 @@ public class MySqlDataSourceOptions {
                     .defaultValue(true)
                     .withDescription(
                             "Whether send schema change events, by default is true. If set to false, the schema changes will not be sent.");
+
+    @Experimental
+    public static final ConfigOption<String> TABLE_EXCLUDE_LIST =
```
Comment (PatrickRen) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/org/apache/flink/cdc/connectors/mysql/source/MySqlDataSourceOptions.java:236]:
```suggestion
    public static final ConfigOption<String> TABLES_EXCLUDE =
```

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/test/java/org/apache/flink/cdc/connectors/mysql/source/MySqlDataSourceFactoryTest.java [hunk 6]
```diff
@@ -79,6 +80,27 @@ public void testNoMatchedTable() {
                 .hasMessageContaining("Cannot find any table by the option 'tables' = " + tables);
     }
 
+    @Test
```
Comment (PatrickRen) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/test/java/org/apache/flink/cdc/connectors/mysql/source/MySqlDataSourceFactoryTest.java:83]: We still need a test case for validating if excluding part of tables (not all) works as expected

File: docs/content.zh/docs/connectors/mysql.md [hunk 7]
```diff
@@ -107,6 +107,14 @@ pipeline:
           需要注意的是，点号（.）被视为数据库和表名的分隔符。 如果需要在正则表达式中使用点（.）来匹配任何字符，必须使用反斜杠对点进行转义。<br>
           例如，db0.\.*, db1.user_table_[0-9]+, db[1-2].[app|web]order_\.*</td>
     </tr>
+    <tr>
+      <td>tables.exclude</td>
+      <td>optional</td>
+      <td style="word-wrap: break-word;">(none)</td>
+      <td>String</td>
+      <td>需要排除的 MySQL 数据库的表名,参数会在tables参数后发生排除作用。表名支持正则表达式，以排除满足正则表达式的多个表。<br>
```
Comment (PatrickRen) [inline docs/content.zh/docs/connectors/mysql.md:115]: nit:
```suggestion
      <td>需要排除的 MySQL 数据库的表名，参数会在tables参数后发生排除作用。表名支持正则表达式，以排除满足正则表达式的多个表。<br>
```

**Review by PatrickRen — COMMENTED** [review]
@shiyiky Thanks for the update! I left some comments.

**Comment by shiyiky** [comment]
@PatrickRen  Thank you for your suggestion and I have done it. Please review it if u have free time.

**Review by PatrickRen — APPROVED** [review]
@shiyiky Thanks for the update! LGTM

**Pull request merged**